tsdb/show_tag_keys.go

```diff
@@ -0,0 +1,315 @@
+package tsdb
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/influxdb/influxdb/influxql"
+	"github.com/influxdb/influxdb/models"
+)
+
+// ShowTagKeysExecutor implements the Executor interface for a SHOW MEASUREMENTS statement.
+type ShowTagKeysExecutor struct {
+	stmt      *influxql.ShowTagKeysStatement
+	mappers   []Mapper
+	chunkSize int
+}
+
+// NewShowTagKeysExecutor returns a new ShowTagKeysExecutor.
+func NewShowTagKeysExecutor(stmt *influxql.ShowTagKeysStatement, mappers []Mapper, chunkSize int) *ShowTagKeysExecutor {
+	return &ShowTagKeysExecutor{
+		stmt:      stmt,
+		mappers:   mappers,
+		chunkSize: chunkSize,
+	}
+}
+
+// Execute begins execution of the query and returns a channel to receive rows.
+func (e *ShowTagKeysExecutor) Execute() <-chan *models.Row {
+	// Create output channel and stream data in a separate goroutine.
+	out := make(chan *models.Row, 0)
+
+	// It's important that all resources are released when execution completes.
+	defer e.close()
+
+	go func() {
+		defer close(out)
+		// Open the mappers.
+		for _, m := range e.mappers {
+			if err := m.Open(); err != nil {
+				out <- &models.Row{Err: err}
+				return
+			}
+		}
+
+		// Create a map of measurement to tags keys.
+		set := map[string]map[string]struct{}{}
+		// Iterate through mappers collecting measurement names.
+		for _, m := range e.mappers {
+			// Read all data from the mapper.
+			for {
+				c, err := m.NextChunk()
+				if err != nil {
+					out <- &models.Row{Err: err}
+					return
+				} else if c == nil {
+					// Mapper has been drained.
+					break
+				}
+
+				// Convert the mapper chunk to an array of measurements with tag keys.
+				mtks, ok := c.(MeasurementsTagKeys)
+				if !ok {
+					out <- &models.Row{Err: fmt.Errorf("show tag keys mapper returned invalid type: %T", c)}
+					return
+				}
+
+				// Merge mapper chunk with previous mapper outputs.
+				for _, mm := range mtks {
+					for _, key := range mm.TagKeys {
+						if set[mm.Measurement] == nil {
+							set[mm.Measurement] = map[string]struct{}{}
+						}
+						set[mm.Measurement][key] = struct{}{}
+					}
+				}
+			}
+		}
+
+		// All mappers are drained.
+
+		// Convert the set into an array of measurements and their tag keys.
+		mstks := make(MeasurementsTagKeys, 0)
+		for mm, tks := range set {
+			mtks := &MeasurementTagKeys{Measurement: mm}
+			for tk := range tks {
+				mtks.TagKeys = append(mtks.TagKeys, tk)
+			}
+			sort.Strings(mtks.TagKeys)
+			mstks = append(mstks, mtks)
+		}
+		// Sort by measurement name.
+		sort.Sort(mstks)
+
+		slim, soff := limitAndOffset(e.stmt.SLimit, e.stmt.SOffset, len(mstks))
```

Comment:
I don't know how much it matters with this call -- and I think the rest of our code suffers from this problem too -- but it looks like all the data is pulled back from the mappers, and then the `LIMIT` and `SLIMIT` is applied. We need to fix this so that the mappers stop providing data once the limits have been reach.

Another time, I know there is an issue open about it right now.